### PR TITLE
Restructure CSRF

### PR DESCRIPTION
**Pull request by aronerben**
https://github.com/oxidizing/sihl/issues/400
https://github.com/oxidizing/sihl/issues/408

### CODE REVIEW

**Comment by aronerben** [comment]
@jerben I've added many new tests to cover more cases than before and ensure more refactor safety.

File: sihl/src/web_csrf.ml [hunk 1]
```diff
@@ -82,108 +37,65 @@ let decrypt_with_salt ~salted_cipher ~salt_length =
...
-    match token, is_safe with
-    | None, true ->
-      (* Don't check for CSRF token in safe requests *)
-      (match stored_secret with
-      | Some secret ->
-        let token = secret_to_token secret in
-        let req = set token req in
+    if Core_configuration.is_development ()
+       && Option.value
+            (Core_configuration.read_bool "NO_CSRF_BYPASS")
```
Comment (jerben) [inline sihl/src/web_csrf.ml:52]: @aronerben Should we maybe use something like `FORCE_CSRF_CHECK` to avoid the double negative?

**Review by jerben — CHANGES_REQUESTED** [review]
@aronerben The implementation looks very clean and simple, nice! Also thanks for the thorough tests. One thing: We can remove the `Sihl.Cleaner.clean_all` calls in the tests since the implementation is stateless.

**Comment by jerben** [comment]
@aronerben Could you check how many rounds are used? We might want to pull the rounds from a central place that we can update every now and then depending what the current best practices are.

File: sihl/src/web_csrf.ml (hunk 1 above)
Comment (aronerben) [reply sihl/src/web_csrf.ml]: Yea, you're right, done.

**Comment by aronerben** [comment]
> @aronerben The implementation looks very clean and simple, nice! Also thanks for the thorough tests. One thing: We can remove the `Sihl.Cleaner.clean_all` calls in the tests since the implementation is stateless.

Thanks, done.

**Comment by aronerben** [comment]
> @aronerben Could you check how many rounds are used? We might want to pull the rounds from a central place that we can update every now and then depending what the current best practices are.

`Bcrypt` requires atleast 4 rounds ands adds a salt. This adds some overhead and might impact performance. Since we rotate the CSRF token on every request now, it's smarter to hash with one round and with a static secret to replace the salt. I've done that accordingly.

**Review by jerben — APPROVED** [review]

**Pull request merged**